src/k2/httpproxy/HTTPProxy.cpp

```diff
@@ -316,7 +235,45 @@ HTTPProxy::_handleRead(shd::ReadRequest&& request) {
 seastar::future<std::tuple<sh::Status, shd::QueryResponse>>
 HTTPProxy::_handleQuery(shd::QueryRequest&& request) {
     K2LOG_D(log::httpproxy, "Received query request {}", request);
-    return MakeHTTPResponse<shd::QueryResponse>(sh::Statuses::S501_Not_Implemented("query not implemented"), shd::QueryResponse{});
+    auto iter = _txns.find(request.timestamp);
+    if (iter == _txns.end()) {
+        return MakeHTTPResponse<shd::QueryResponse>(Txn_S410_Gone, shd::QueryResponse{});
+    }
+    auto queryIter = iter->second.queries.find(request.queryId);
+    if (queryIter ==iter->second.queries.end()) {
+        K2LOG_W(log::httpproxy, "Query not found, txn: {} query: {}", request.timestamp, request.queryId);
+        return MakeHTTPResponse<shd::QueryResponse>(Query_S410_Gone, shd::QueryResponse{});
+    }
+
+    return iter->second.handle.query(queryIter->second)
+    .then([this, request=std::move(request)](QueryResult&& result) {
+        if(!result.status.is2xxOK()) {
+            return MakeHTTPResponse<shd::QueryResponse>(sh::Status{.code = result.status.code, .message = result.status.message}, shd::QueryResponse{});
+        }
+        std::vector<shd::SKVRecord::Storage> records;
+        records.reserve(result.records.size());
+        for (auto& k2record: result.records) {
+            sh::String collectionName(k2record.collectionName);
+            // k2 schema is already populated by query api, get corresponding shd schema from cache
+            auto shSChema = getSchemaFromCache(collectionName, k2record.schema);
+            auto rec = _buildSHDRecord(k2record, collectionName, shSChema);
+            records.push_back(std::move(rec.getStorage()));
+        }
+        if (auto iter = _txns.find(request.timestamp); iter == _txns.end()) {
+            return MakeHTTPResponse<shd::QueryResponse>(Txn_S410_Gone, shd::QueryResponse{});
+        } else {
+            if (auto queryIter = iter->second.queries.find(request.queryId); queryIter ==iter->second.queries.end()) {
+                K2LOG_W(log::httpproxy, "Query not found, txn: {} query: {}", request.timestamp, request.queryId);
+                return MakeHTTPResponse<shd::QueryResponse>(Query_S410_Gone, shd::QueryResponse{});
+            } else {
+                if (queryIter->second.isDone()) {
+                    iter->second.queries.erase(request.queryId);
+                }
+                return MakeHTTPResponse<shd::QueryResponse>(sh::Status{.code=result.status.code, .message=result.status.message},
+                    shd::QueryResponse{.done = queryIter->second.isDone(), .records = std::move(records)});
```

Comment:
Is it safe to access queryIter here if the erase happens above?

Reply:
Good catch! I mistakenly refactored this way in previous update to reduce a line of code.